### PR TITLE
Fix the Flatpak

### PR DESCRIPTION
**Pull request by AsciiWolf**
It now works fine and is probably ready to be submitted to Flathub. :-)

As far as I know, the reproducible builds rule has some exceptions for apps without proper releases so it hopefully should not be a problem (if a valid commit hash is included).

Feel free to let me know if you need any help with publishing the Flatpak on Flathub.

### CODE REVIEW

**Comment by mattiasb** [comment]
Cool, thanks! As I'm reading this it's way past bedtime here and I'm busy for a couple of days. If I haven't gotten to review this by say Monday evening, could you poke me? It would be neat to get IOQuake3 on flathub!

**Comment by AsciiWolf** [comment]
@mattiasb A little poke about this PR and the Flathub submission. :-)

By the way, one of the Flathub submission criteria is that "any additional patches or files should be submitted to the upstream projects concerned". This in our case means the AppStream metadata file. I can make the upstream Pull Request if you want.

**Comment by mattiasb** [comment]
Cool! *Hopefully* I'll get to this tomorrow morning.

I *might* ask you to just fork after a quick review tomorrow to be honest though. Life is busy as it is. :) Again, I would *love* to see this on flathub even though I only play Q3 a few times a year during Eurovision Song Contest (we call it "No Mello¹ for old men!" 😂). So don't take a request to fork as a discouragement!

1: Mello is Swedish slang for ESC

**Comment by AsciiWolf** [comment]
No worries. :-) I have submitted the AppStream file (with slight upstream-specific modifications) into upstream: https://github.com/ioquake/ioq3/pull/607

Feel free to let me know when you have enough time to properly work on submitting the Flatpak to Flathub, no need to rush.

**Comment by mattiasb** [comment]
Just remembered this now. Will do a review. :)

File: README.md [hunk 1]
```diff
@@ -13,11 +13,8 @@ $ flatpak-builder build --install --user ./org.ioquake3.IOQuake3.yaml
 Definitely!
 
 I just need to:
- - … [add an appstream file][1]
- - … [convince upstream to make releases][2]
- - … [ensure I don't keep any crap around][3]
- - … add a flathub.json file (I believe).
+ - … [convince upstream to make releases][1]
+ - … [ensure I don't keep any crap around][2]
```
Comment (mattiasb) [inline README.md:17]: If I remember correctly this is about putting files inside the Flatpak that's not needed. I don't think it's *that* important but might as well keep it here. Could turn these into issues later.

File: README.md [hunk 2]
```diff
@@ -13,11 +13,8 @@ $ flatpak-builder build --install --user ./org.ioquake3.IOQuake3.yaml
 Definitely!
 
 I just need to:
- - … [add an appstream file][1]
- - … [convince upstream to make releases][2]
- - … [ensure I don't keep any crap around][3]
- - … add a flathub.json file (I believe).
+ - … [convince upstream to make releases][1]
```
Comment (mattiasb) [inline README.md:16]: Hm, did you get the state of this? Like can we submit if there's no release?

File: ioquake3 [hunk 3]
```diff
@@ -6,24 +6,31 @@
 function error-message {
     cat << EOM
 ioquake3 needs the official Quake 3 .pk3-files to function
-properly (specifically pak0.pk3 … pak8.pk3).
+properly (specifically baseq3/pak0.pk3 … baseq3/pak8.pk3 and
+optionally also missionpack/pak0.pk3 … missionpack/pak3.pk3).
```
Comment (mattiasb) [inline ioquake3:10]: Ah nice!

File: ioquake3 [hunk 4]
```diff
@@ -6,24 +6,31 @@
...
+Once you've obtained all the required files, copy them to:
+
+    ~/.var/app/org.ioquake3.IOQuake3/.q3a/
 
 … and restart ioquake3
 EOM
 }
 
 # Ensure ioquake3 reaches our pak-files and configs.
-ln -ns "${XDG_DATA_HOME}" "${HOME}/.q3a"
```
Comment (mattiasb) [inline ioquake3:23]: Hm. You will have to explain this one I think.

Basically Quake 3 is far too old for the XDG directory spec (and IOQuake probably hasn't bothered moving with the times here), but if we want the `.q3a` directory reachable from *outside* the Flatpak this directory probably needs to map somewhere inside an XDG directory I think. Since the directory contains both data and config, I think `XDG_DATA_HOME` is the way to go.

It could make sense to do this instead though:
```
ln -ns "${XDG_DATA_HOME}/q3a" "${HOME}/.q3a"
```

such that we don't take over the whole data directory. But not sure if that matters.

File: ioquake3 [hunk 5]
```diff
@@ -6,24 +6,31 @@
...
+[ ! -f "${HOME}/.q3a/baseq3" ] && mkdir -p "${HOME}/.q3a/baseq3"
+[ ! -f "${HOME}/.q3a/missionpack" ] && mkdir -p "${HOME}/.q3a/missionpack"
```
Comment (mattiasb) [inline ioquake3]: I had to read this a few times. Is the reason for the `-f` that you want to avoid `mkdir` failing if there exists a *file* (not a directory) named `"${HOME}/.q3a/baseq3`?

If you want to check for a directory you would do it with `-d` (I might be mansplaining now, sorry in advance if so). 

Anyhow `mkdir -p` is idempotent, meaning that a second run is a no-op. So just
```shell
mkdir -p  "${HOME}"/.q3a/{baseq3,missionpack}
```
*should* be enough and is what I would do.

If we want to handle edge cases we could instead do:
```shell
if ! out="$(mkdir -p /etc/{lol,wee} 2>&1)"; then
    zenity --error --text "Failed to create directories!\n\n${out}"
    exit 2
fi
```

File: org.ioquake3.IOQuake3.yaml [hunk 6]
```diff
@@ -1,23 +1,23 @@
...
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --persist=.q3a
 
 modules:
   - name: ioquake3
     buildsystem: simple
     build-commands:
-      - BUILD_SERVER=0 COPYDIR=/app make copyfiles
+      - BUILD_SERVER=0 COPYDIR=/app/share/ioquake3 make -j $FLATPAK_BUILDER_N_JOBS copyfiles
```
Comment (mattiasb) [inline org.ioquake3.IOQuake3.yaml:20]: I don't understand the `COPYDIR` change. Could you elaborate?

File: org.ioquake3.IOQuake3.yaml [hunk 7]
```diff
@@ -1,23 +1,23 @@
...
 finish-args:
-  - --socket=pulseaudio
-  - --socket=x11
-  - --share=network
   - --device=dri
+  - --share=network
+  - --socket=pulseaudio
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --persist=.q3a
```
Comment (mattiasb) [inline org.ioquake3.IOQuake3.yaml:14]: In general try to make the diff as small as possible. It helps the reviewer to see what has changed. The `--socket=pulseaudio` and `--share=network` wasn't removed for example and I had to scan further below to see that. One could ofcourse do a resort of the arguments (I'm one of few developers who likes that kind of commit) but then it should be a separate commit. Again for review purposes.

Anyways, these comments are for sure no biggie at all for this project. I just had to say it because it might be useful in the future. :)

So I think the real change here is:
```diff
-  - --socket=x11
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --persist=.q3a
```

And regarding that I think the `  - --persist=.q3a` change can go if you revert the `XDG_DATA_HOME` change.

File: org.ioquake3.IOQuake3.metainfo.xml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    </screenshot>
+  </screenshots>
+  <developer_name>The ioquake Group</developer_name>
+  <releases>
+    <release version="1.36_GIT" date="2023-07-06"/>
+  </releases>
+  <content_rating type="oars-1.1">
+    <content_attribute id="violence-cartoon">intense</content_attribute>
+    <content_attribute id="violence-fantasy">intense</content_attribute>
+    <content_attribute id="violence-bloodshed">mild</content_attribute>
```
Comment (mattiasb) [inline org.ioquake3.IOQuake3.metainfo.xml]: I actually don't remember this since I have no issues personally with blood. Could you double check that the amount of bloodshed is really "mild"? I *do* remember Quake 1 though and that's definitely not "mild" 😂

File: org.ioquake3.IOQuake3.yaml [hunk 9]
```diff
@@ -1,23 +1,23 @@
----
```
Comment (mattiasb) [inline org.ioquake3.IOQuake3.yaml:1]: Why did you remove this?

File: org.ioquake3.IOQuake3.yaml [hunk 10]
```diff
@@ -1,23 +1,23 @@
...
 app-id: org.ioquake3.IOQuake3
 runtime: org.freedesktop.Platform
-runtime-version: '20.08'
+runtime-version: '22.08'
 sdk: org.freedesktop.Sdk
-branch: stable
```
Comment (mattiasb) [inline org.ioquake3.IOQuake3.yaml:6]: Hm looked it up and indeed setting `branch` specifically is frowned upon. Maybe set `default-branch` to `stable` instead?

**Comment by mattiasb** [comment]
I think my main gripe is that the PR contains a bunch of changes that feels "snuck in". Like the `XDG_DATA_HOME` change for example. It would be easier to review if the logical changes were split into different commits and the commits having commit messages stating why the changes are made.

With that said. This isn't a very important repo for me in that sense so I'm fine with just having it all squashed into a single commit. I want a commit message though stating what the commit does and why (feel free to do `commit --amend` and force pushing!).

File: ioquake3 (hunk 5 above)
Comment (mattiasb) [reply ioquake3]: Oh and prefer `if this; then that; fi` instead. I used to rely on `&&` as well but there are some pitfalls (that i don't remember off hand).

**Comment by mattiasb** [comment]
Oh. And run `yamllint` on the manifest and `shellcheck` on the wrapper script as well. I haven't had time to play with this locally yet so haven't looked for lint errors. (For a proper project those should be caught by CI but yeah).

**Comment by mattiasb** [comment]
Ah, the actual review was in pending state. :D There now the comments should be up!

File: README.md [hunk 11]
```diff
@@ -13,11 +13,8 @@ $ flatpak-builder build --install --user ./org.ioquake3.IOQuake3.yaml
 Definitely!
 
 I just need to:
- - … [add an appstream file][1]
- - … [convince upstream to make releases][2]
- - … [ensure I don't keep any crap around][3]
- - … add a flathub.json file (I believe).
```
Comment (mattiasb) [inline README.md:19]: Does ioquake3 work on all arches supported by Flathub? I think that was my reason for a flathub.json.

**Comment by AsciiWolf** [comment]
> I think my main gripe is that the PR contains a bunch of changes that feels "snuck in". Like the `XDG_DATA_HOME` change for example. It would be easier to review if the logical changes were split into different commits and the commits having commit messages stating why the changes are made.
> 
> With that said. This isn't a very important repo for me in that sense so I'm fine with just having it all squashed into a single commit. I want a commit message though stating what the commit does and why (feel free to do `commit --amend` and force pushing!).

Thanks for the review! I agree that the all-in-one commit can be confusing. I will edit the commit message to list all the changes I have made.

**Comment by AsciiWolf** [comment]
> Oh. And run `yamllint` on the manifest and `shellcheck` on the wrapper script as well. I haven't had time to play with this locally yet so haven't looked for lint errors. (For a proper project those should be caught by CI but yeah).

```
$ yamllint org.ioquake3.IOQuake3.yaml
org.ioquake3.IOQuake3.yaml
  1:1       warning  missing document start "---"  (document-start)
  20:81     error    line too long (92 > 80 characters)  (line-length)
  48:81     error    line too long (81 > 80 characters)  (line-length)
```

The warning is incorrect. Flatpak manifest yamls do not start with "---". The two errors can be waived, it's just a few extra characters on one line and they do not make the file look bad.

```
$ shellcheck ioquake3
In ioquake3 line 42:
/app/share/ioquake3/ioquake3.* "$@"
^----------------------------^ SC2211 (warning): This is a glob used as a command name. Was it supposed to be in ${..}, array, or is it missing quoting?

For more information:
  https://www.shellcheck.net/wiki/SC2211 -- This is a glob used as a command ...
```
The launch command is correct. I have replaced the hardcoded arch with an asterisk to make it work on other arches as well. There are no downsides to this easy solution since there is no other file named ioquake3.* in that directory.

```
$ appstreamcli validate org.ioquake3.IOQuake3.metainfo.xml

✔ Validation was successful: pedantic: 3
```

**Comment by AsciiWolf** [comment]
```
$ appstreamcli validate org.ioquake3.IOQuake3.metainfo.xml

✔ Validation was successful: pedantic: 3
```

File: README.md (hunk 1 above)
Comment (AsciiWolf) [reply README.md:17]: I have decided to keep this point in your list, but feel free to remove it. I did already check this and found no leftover files that are not needed in the Flatpak.

File: README.md (hunk 2 above)
Comment (AsciiWolf) [reply README.md:16]: As I said, as far as I know, there is an exception for such apps on Flathub. I have already did this on my Flathub apps.

File: README.md (hunk 11 above)
Comment (AsciiWolf) [reply README.md:19]: The only other active arch at the moment is aarch64 and since my OpenArena Flatpak that also uses ioquake3 works fine there, ioquake3 itself should also work with no issues. I have already modified the launcher script and removed the hardcoded x86_64 binary arch.

File: ioquake3 (hunk 3 above)
Comment (AsciiWolf) [reply ioquake3:10]: :-)

File: ioquake3 (hunk 4 above)
Comment (AsciiWolf) [reply ioquake3:23]: Creating a symlink every time the Flatpak is launcher is really not ideal solution, at least in my opinion. For such cases,  the - `--persist=foo` is the way to go.

And if we use persist, we cannot do `ln -ns "${XDG_DATA_HOME}" "${HOME}/.q3a"`, because the directory already exists (gets created by the Flatpak itself).

See for example the OpenArena Flatpak: https://github.com/flathub/ws.openarena.OpenArena/blob/master/ws.openarena.OpenArena.json#L13

Persisting directories are perfectly accessible outside of Flatpak, for example in our case:

~/.var/app/org.ioquake3.IOQuake3/.q3a/

File: ioquake3 (hunk 5 above)
Comment (AsciiWolf) [reply ioquake3]: You are right. :-) I completely forgot about `mkdir -p` not creating or touching the directory if it already exists. I will fix it, thanks!

File: org.ioquake3.IOQuake3.yaml (hunk 9 above)
Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.yaml:1]: Flatpak manifest yamls do not start with "---".

File: org.ioquake3.IOQuake3.yaml (hunk 10 above)
Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.yaml:6]: Manually specified branch is not needed and not used anymore (at least on Flathub). And from my experience, it can cause weird issues.

Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.yaml:6]: See for example one of my newer Flatpaks:

https://github.com/flathub/net.worldofpadman.WoP/blob/master/net.worldofpadman.WoP.yaml

File: org.ioquake3.IOQuake3.yaml (hunk 7 above)
Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.yaml:14]: Yeah, I have also fixed the finish-args positions to make it look a little bit better. Sorry about that. :-)

I have just changed the `x11` socket to `fallback-x11`, added the wayland socket and added persisting `.q3a` directory. As I stated above, reverting back to XDG_DATA_HOME and symlink creation on every startup is not a good idea, at least on my opinion (as someone who maintains 30+ apps on Flathub for last ~5 years).

File: org.ioquake3.IOQuake3.yaml (hunk 6 above)
Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.yaml:20]: Sure. You were installing ioquake3 directly into the root (`/app`) filesystem. This is not how the Flatpak file system structure works. I have changed it to a correct (`/app/share/ioquake3`) path.

File: org.ioquake3.IOQuake3.metainfo.xml (hunk 8 above)
Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.metainfo.xml]: Well, here is a screenshot from the [AppStream metadata generator](https://hughsie.github.io/oars/generate.html):

bloodshed

To be honest, I am not sure, but blood effects of Quake 3 Arena do not look much realistic to me, at least from today's standpoint. It is the same thing as the "Realistic Violence" rating in OARS that I set to "None" and kept only the "Cartoon Violence" and "Fantasy Violence" ones.

File: README.md [hunk 12]
```diff
@@ -13,11 +13,8 @@ $ flatpak-builder build --install --user ./org.ioquake3.IOQuake3.yaml
 Definitely!
 
 I just need to:
```
Comment (AsciiWolf) [inline README.md:15]: After merging my fixes and submitting the Flatpak as a PR to Flathub, there is one more thing we need to do:

`- … [ask upstream about their interest in co-maintaining the Flatpak]`

In other words, just [create a ticket here](https://github.com/ioquake/ioq3/issues) about the Flatpak and upstream push rights to its repo (as described in the new PR template on Flathub).

File: ioquake3 (hunk 5 above)
Comment (AsciiWolf) [reply ioquake3]: Done. :-)

File: README.md (hunk 12 above)
Comment (AsciiWolf) [reply README.md:15]: Also, when submitting the final PR to Flathub, it would be good to remove the README file or replace it with something useful to end-users.

**Comment by AsciiWolf** [comment]
> I want a commit message though stating what the commit does and why (feel free to do `commit --amend` and force pushing!).

Done. ;-) Feel free to re-review.

File: org.ioquake3.IOQuake3.metainfo.xml (hunk 8 above)
Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.metainfo.xml]: This is how the OpenArena OARS metadata look like:

```
    <content_attribute id="violence-fantasy">moderate</content_attribute>
    <content_attribute id="violence-bloodshed">mild</content_attribute>
    <content_attribute id="social-chat">intense</content_attribute>
```

Q3A is a little bit more realistic, so the intense _violence-fantasy_ make sense, but I think I will keep the _violence-bloodshed_ on mild. Also, I will probably remove the _violence-cartoon_.

Comment (AsciiWolf) [reply org.ioquake3.IOQuake3.metainfo.xml]: I thought about it again and decided to change the _violence-bloodshed_ to intense. And also to remove the _violence-cartoon_.

**Comment by AsciiWolf** [comment]
Superseded by #2.
